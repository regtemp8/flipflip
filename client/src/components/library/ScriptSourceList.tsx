import React, {
  ChangeEvent,
  useEffect,
  useState,
  useRef,
  useCallback
} from 'react'
import { SortableContainer, SortableElement } from 'react-sortable-hoc'
import AutoSizer from 'react-virtualized-auto-sizer'
import { FixedSizeList } from 'react-window'

import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  List,
  type Theme,
  Typography
} from '@mui/material'

import { makeStyles } from 'tss-react/mui'

import ScriptSourceListItem from './ScriptSourceListItem'
import SceneSelect from '../configGroups/SceneSelect'
import ScriptOptions from './ScriptOptions'
import { SP } from 'flipflip-common'
import { useAppDispatch, useAppSelector } from '../../store/hooks'
import {
  systemMessage,
  setScriptYOffset,
  setScriptsRemoveOne
} from '../../store/app/slice'
import { swapScripts, setScriptsEditUrl } from '../../store/app/thunks'
import { playScript } from '../../store/scene/thunks'
import {
  selectAppScriptYOffset,
  selectAppSpecialMode,
  selectUndefined
} from '../../store/app/selectors'
import { selectCaptionScriptUrl } from '../../store/captionScript/selectors'
import flipflip from '../../FlipFlipService'

const useStyles = makeStyles()((theme: Theme) => ({
  emptyMessage: {
    textAlign: 'center',
    marginTop: '25%'
  },
  emptyMessage2: {
    textAlign: 'center'
  },
  marginRight: {
    marginRight: theme.spacing(1)
  },
  arrow: {
    position: 'absolute',
    bottom: 20,
    right: 35,
    fontSize: 220,
    transform: 'rotateY(0deg) rotate(45deg)'
  },
  arrowMessage: {
    position: 'absolute',
    bottom: 260,
    right: 160
  },
  scenePick: {
    height: 410
  }
}))

interface SortableVirtualListProps {
  height: number
  width: number
  yOffset: number
  sources: number[]
}

export interface ScriptSourceListProps {
  showHelp: boolean
  sources: number[]
  selected: number[]
  onUpdateSelected: (selected: number[]) => void
}

function ScriptSourceList(props: ScriptSourceListProps) {
  const dispatch = useAppDispatch()
  const [sourceOptions, setSourceOptions] = useState<number>()
  const [lastSelected, setLastSelected] = useState<number>()
  const [isEditing, setIsEditing] = useState(-1)
  const [deleteDialog, setDeleteDialog] = useState<number>()
  const [beginPlay, setBeginPlay] = useState<number>()
  const [playWithScene, setPlayWithScene] = useState<number>()

  const deleteDialogSelector =
    deleteDialog != null
      ? selectCaptionScriptUrl(deleteDialog)
      : selectUndefined
  const deleteDialogURL = useAppSelector(deleteDialogSelector)
  const beginPlaySelector =
    beginPlay != null ? selectCaptionScriptUrl(beginPlay) : selectUndefined
  const beginPlayURL = useAppSelector(beginPlaySelector)
  const firstSourceURL = useAppSelector(
    selectCaptionScriptUrl(props.sources.length > 0 ? props.sources[0] : -1)
  )
  const specialMode = useAppSelector(selectAppSpecialMode())
  const yOffset = useAppSelector(selectAppScriptYOffset())

  const _shiftDown = useRef<boolean>()
  const _lastChecked = useRef<number>()

  const savePosition = useCallback(() => {
    const sortableList = document.getElementById('sortable-list')
    if (sortableList) {
      const scrollElement = sortableList.firstElementChild
      const scrollTop = scrollElement ? scrollElement.scrollTop : 0
      dispatch(setScriptYOffset(scrollTop))
    }
  }, [dispatch])

  useEffect(() => {
    const onKeyDown = (e: KeyboardEvent) => {
      if (e.key === 'Shift') _shiftDown.current = true
    }
    const onKeyUp = (e: KeyboardEvent) => {
      if (e.key === 'Shift') _shiftDown.current = false
    }

    window.addEventListener('keydown', onKeyDown, false)
    window.addEventListener('keyup', onKeyUp, false)
    _shiftDown.current = false
    return () => {
      window.removeEventListener('keydown', onKeyDown)
      window.removeEventListener('keyup', onKeyUp)
      _shiftDown.current = undefined
      _lastChecked.current = undefined
      savePosition()
    }
  }, [savePosition])

  useEffect(() => {
    if (firstSourceURL === '') {
      setIsEditing(props.sources[0])
    }
  }, [props.sources, firstSourceURL])

  const onSortEnd = ({
    oldIndex,
    newIndex
  }: {
    oldIndex: number
    newIndex: number
  }) => {
    const oldSourceID = props.sources[oldIndex]
    const newSourceID = props.sources[newIndex]
    dispatch(swapScripts(oldSourceID, newSourceID))
  }

  const clearLastSelected = () => {
    if (!sourceOptions) {
      setLastSelected(undefined)
    }
  }

  const onSourceOptions = (scriptID: number) => {
    setSourceOptions(scriptID)
    setLastSelected(scriptID)
  }

  const onCloseSourceOptions = () => {
    setSourceOptions(undefined)
  }

  const onDelete = (scriptID: number) => {
    setDeleteDialog(scriptID)
  }

  const onCloseDeleteDialog = () => {
    setDeleteDialog(undefined)
  }

  const onFinishDelete = async () => {
    await flipflip().api.unlink(deleteDialogURL as string)
    onRemove(deleteDialog as number)
    onCloseDeleteDialog()
  }

  const onRemove = (scriptID: number) => {
    props.onUpdateSelected(props.selected.filter((id) => id !== scriptID))
    dispatch(setScriptsRemoveOne(scriptID))
  }

  const onToggleSelect = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.currentTarget.value
    const scriptID = Number(value)
    if (specialMode === SP.selectSingle) {
      props.onUpdateSelected([scriptID])
    } else {
      const newSelected = Array.from(props.selected)
      const index = newSelected.indexOf(scriptID)
      if (index !== -1) {
        newSelected.splice(index, 1)
      } else {
        if (
          _lastChecked.current &&
          props.sources.includes(_lastChecked.current) &&
          _shiftDown.current
        ) {
          let start = false
          for (const id of props.sources) {
            if (start && (id === scriptID || id === _lastChecked.current)) {
              break
            }
            if (start) {
              newSelected.push(id)
            }
            if (!start && (id === scriptID || id === _lastChecked.current)) {
              start = true
            }
          }
        }
        newSelected.push(scriptID)
      }
      _lastChecked.current = scriptID
      props.onUpdateSelected(newSelected)
    }
  }

  const onStartEdit = (id: number) => {
    setIsEditing(id)
  }

  const onEndEdit = (newURL: string) => {
    dispatch(setScriptsEditUrl(isEditing, newURL))
    setIsEditing(-1)
  }

  const onPlay = (scriptID: number) => {
    setBeginPlay(scriptID)
    setPlayWithScene(0)
  }

  const onClosePlayDialog = () => {
    setBeginPlay(undefined)
    setPlayWithScene(undefined)
  }

  const onFinishPlay = () => {
    savePosition()
    try {
      const scriptID = beginPlay as number
      const sceneID = playWithScene as number
      dispatch(playScript(scriptID, sceneID, props.sources))
    } catch (e) {
      dispatch(
        systemMessage('The source ' + beginPlayURL + " isn't in your Library")
      )
    }
  }

  const onChangeScene = (sceneID: number) => {
    setPlayWithScene(sceneID)
  }

  function VirtualList(props: {
    height: number
    width: number
    yOffset: number
    sources: any[]
  }) {
    const { height, width, yOffset, sources } = props

    return (
      <FixedSizeList
        height={height}
        width={width}
        initialScrollOffset={yOffset}
        itemSize={56}
        itemCount={sources.length}
        itemData={sources}
        itemKey={(index: number, data: any) => data[index].id}
        overscanCount={10}
      >
        {Row}
      </FixedSizeList>
    )
  }

  function Row(props: any) {
    const { index } = props
    const SortableItem = SortableElement<{ index: any; value: any }>(
      ({ value }: { value: { index: number; style: any; data: any[] } }) => {
        const index = value.index
        const scriptID: number = value.data[index]
        return (
          <ScriptSourceListItem
            key={index}
            checked={specialMode ? props.selected.includes(scriptID) : false}
            index={index}
            isEditing={isEditing}
            lastSelected={scriptID === lastSelected}
            scriptID={scriptID}
            style={value.style}
            onDelete={onDelete}
            onEndEdit={onEndEdit}
            onPlay={onPlay}
            onRemove={onRemove}
            onSourceOptions={onSourceOptions}
            onStartEdit={onStartEdit}
            onToggleSelect={onToggleSelect}
            savePosition={savePosition}
          />
        )
      }
    )

    return <SortableItem index={index} value={props} />
  }

  const { classes } = useStyles()
  if (props.sources.length === 0) {
    return (
      <React.Fragment>
        <Typography
          component="h1"
          variant="h3"
          color="inherit"
          noWrap
          className={classes.emptyMessage}
        >
          乁( ◔ ౪◔)「
        </Typography>
        <Typography
          component="h1"
          variant="h4"
          color="inherit"
          noWrap
          className={classes.emptyMessage2}
        >
          Nothing here
        </Typography>
        {props.showHelp && (
          <React.Fragment>
            <Typography
              component="h1"
              variant="h6"
              color="inherit"
              noWrap
              className={classes.arrowMessage}
            >
              Add new scripts
            </Typography>
            <div className={classes.arrow}>→</div>
          </React.Fragment>
        )}
      </React.Fragment>
    )
  }

  const SortableVirtualList =
    SortableContainer<SortableVirtualListProps>(VirtualList)
  return (
    <React.Fragment>
      <AutoSizer>
        {({ height, width }: { height: number; width: number }) => (
          <List id="sortable-list" disablePadding onClick={clearLastSelected}>
            <SortableVirtualList
              helperContainer={() =>
                document.getElementById('sortable-list') as HTMLElement
              }
              distance={5}
              height={height - 1}
              width={width}
              yOffset={yOffset}
              sources={props.sources}
              onSortEnd={onSortEnd}
            />
          </List>
        )}
      </AutoSizer>
      {deleteDialog != null && (
        <Dialog
          open={true}
          onClose={onCloseDeleteDialog}
          aria-describedby="delete-description"
        >
          <DialogContent>
            <DialogContentText id="delete-description">
              Are you sure you want to delete {deleteDialogURL}?
            </DialogContentText>
          </DialogContent>
          <DialogActions>
            <Button onClick={onCloseDeleteDialog} color="secondary">
              Cancel
            </Button>
            <Button onClick={onFinishDelete} color="primary">
              Delete
            </Button>
          </DialogActions>
        </Dialog>
      )}
      {sourceOptions != null && (
        <ScriptOptions scriptID={sourceOptions} onDone={onCloseSourceOptions} />
      )}
      {beginPlay != null && (
        <Dialog
          open={true}
          onClose={onClosePlayDialog}
          aria-describedby="play-description"
        >
          <DialogContent className={classes.scenePick}>
            <DialogContentText id="play-description">
              Choose a scene to test with:
            </DialogContentText>
            <SceneSelect
              autoFocus
              menuIsOpen
              value={playWithScene ?? 0}
              onChange={onChangeScene}
            />
          </DialogContent>
          <DialogActions>
            <Button onClick={onClosePlayDialog} color="secondary">
              Cancel
            </Button>
            <Button
              disabled={playWithScene == null || playWithScene === 0}
              onClick={onFinishPlay}
              color="primary"
            >
              Play
            </Button>
          </DialogActions>
        </Dialog>
      )}
    </React.Fragment>
  )
}

;(ScriptSourceList as any).displayName = 'ScriptSourceList'
export default ScriptSourceList
