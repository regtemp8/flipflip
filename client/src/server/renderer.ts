import createCache, { type EmotionCache } from '@emotion/cache'
import { configureStore } from '@reduxjs/toolkit'
import { AppStorage, SystemConstants } from 'flipflip-common'
import createAppReducer from '../store/app/slice'
import createAudioReducer from '../store/audio/slice'
import createCaptionScriptReducer from '../store/captionScript/slice'
import createClipReducer from '../store/clip/slice'
import createLibrarySourceReducer from '../store/librarySource/slice'
import createPlaylistReducer from '../store/playlist/slice'
import createSceneReducer from '../store/scene/slice'
import createSceneGroupReducer from '../store/sceneGroup/slice'
import createTagReducer from '../store/tag/slice'
import playersSlice from '../store/player/slice'
import videoClipperSlice from '../store/videoClipper/slice'
import scenePickerSlice from '../store/scenePicker/slice'
import captionScriptorSlice from '../store/captionScriptor/slice'
import sceneDetailSlice from '../store/sceneDetail/slice'
import audioLibrarySlice from '../store/audioLibrary/slice'
import sourceScraperSlice from '../store/sourceScraper/slice'
import createConstantsReducer from '../store/constants/slice'
import { fromAppStorage } from '../store/app/convert'
import createDisplayReducer from '../store/display/slice'
import createDisplayViewReducer from '../store/displayView/slice'
import createScenePlaylistItemReducer from '../store/scenePlaylistItem/slice'
import createDisplayPlaylistItemReducer from '../store/displayPlaylistItem/slice'

export function createEmotionCache(): EmotionCache {
  return createCache({ key: 'css' })
}

export function createReduxStore(
  appStorage?: AppStorage,
  constants?: SystemConstants
) {
  const state = appStorage != null ? fromAppStorage(appStorage) : undefined
  return configureStore({
    reducer: {
      app: createAppReducer(state?.app),
      audio: createAudioReducer(state?.audio),
      captionScript: createCaptionScriptReducer(state?.captionScript),
      clip: createClipReducer(state?.clip),
      librarySource: createLibrarySourceReducer(state?.librarySource),
      playlist: createPlaylistReducer(state?.playlist),
      displayPlaylistItem: createDisplayPlaylistItemReducer(
        state?.displayPlaylistItem
      ),
      scenePlaylistItem: createScenePlaylistItemReducer(
        state?.scenePlaylistItem
      ),
      scene: createSceneReducer(state?.scene),
      sceneGroup: createSceneGroupReducer(state?.sceneGroup),
      display: createDisplayReducer(state?.display),
      displayView: createDisplayViewReducer(state?.displayView),
      tag: createTagReducer(state?.tag),
      constants: createConstantsReducer(constants),
      // components
      players: playersSlice,
      videoClipper: videoClipperSlice,
      scenePicker: scenePickerSlice,
      captionScriptor: captionScriptorSlice,
      sceneDetail: sceneDetailSlice,
      audioLibrary: audioLibrarySlice,
      sourceScraper: sourceScraperSlice
    }
  })
}
