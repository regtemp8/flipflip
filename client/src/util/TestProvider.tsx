import * as React from 'react'
import { Provider } from 'react-redux'
import { createTheme } from '@mui/material'
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles'
import defaultTheme from 'flipflip-common'

export default class TestProvider extends React.Component {
  readonly props: {
    children: any
    store: any
  }

  disableTransitions = {
    defaultProps: {
      disablePortal: true,
      hideBackdrop: true,
      TransitionComponent: ({ children }) => children
    }
  }

  render() {
    const theme = createTheme({
      ...defaultTheme,
      // disable Transitions when runing tests, throws errors otherwise.
      components: {
        MuiMenu: this.disableTransitions,
        MuiPopover: this.disableTransitions,
        MuiDialog: this.disableTransitions,
        MuiModal: this.disableTransitions
      }
    } as any)
    return (
      <Provider store={this.props.store}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>{this.props.children}</ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    )
  }
}
