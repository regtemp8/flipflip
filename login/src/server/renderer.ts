import createCache, { type EmotionCache } from "@emotion/cache";

export function createEmotionCache(nonce?: string): EmotionCache {
  return createCache({ key: "css", nonce });
}
