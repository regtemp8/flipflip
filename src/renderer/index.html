<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <title>FlipFlip</title>
    <!-- Responsive meta tag -- ensures proper rendering and touch zooming for all devices -->
    <meta name="viewport" content="minimum-scale=1, initial-scale=1, width=device-width, shrink-to-fit=no" />
    <!-- Roboto Font -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap" />
    <!-- Font Icons -->
    <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons" />

    <meta
      http-equiv="Content-Security-Policy"
      content="frame-src https://www.imagefap.com http://www.imagefap.com https://hypno.nimja.com/ http://hypno.nimja.com/; font-src 'self' https://fonts.gstatic.com; style-src 'self' https://fonts.googleapis.com 'unsafe-inline'"
    />
  </head>

  <body>
    <div id="app" class="app"></div>
  </body>
</html>
