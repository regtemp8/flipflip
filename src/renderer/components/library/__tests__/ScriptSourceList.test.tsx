import React from "react";
import { describe, it, expect } from "@jest/globals";
import renderer from "react-test-renderer";
import ScriptSourceList from "../ScriptSourceList";
import TestProvider from "../../../../../test/util/TestProvider";

jest.mock('../ScriptSourceListItem', () => 'ScriptSourceListItem');
jest.mock('../../configGroups/SceneSelect', () => 'SceneSelect');
jest.mock('../ScriptOptions', () => 'ScriptOptions');

describe("ScriptSourceList", () => {
  it("should match snapshot", () => {
    const component = renderer.create(
      <TestProvider>
        <ScriptSourceList
          library={[]}
          scenes={[]}
          showHelp={false}
          sources={[]}
          tutorial={null}
          onEditScript={(source) => {}}
          onPlay={(source, sceneID, displayed) => {}}
          onUpdateLibrary={(fn) => {}}
          onUpdateScript={(script) => {}}
          systemMessage={(message) => {}}
        />
      </TestProvider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
