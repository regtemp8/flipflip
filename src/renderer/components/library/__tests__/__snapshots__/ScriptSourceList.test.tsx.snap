// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScriptSourceList should match snapshot 1`] = `
[
  <h1
    className="MuiTypography-root MuiTypography-h3 MuiTypography-noWrap ScriptSourceList-emptyMessage-1 css-1yk4hom-MuiTypography-root"
  >
    乁( ◔ ౪◔)「
  </h1>,
  <h1
    className="MuiTypography-root MuiTypography-h4 MuiTypography-noWrap ScriptSourceList-emptyMessage2-2 css-1r3v6n7-MuiTypography-root"
  >
    Nothing here
  </h1>,
]
`;
